### PR TITLE
[LG] Inject LG templates as a global expression function 

### PR DESCRIPTION
**Pull request by cosmicshuai**
close: #3652
In this PR, users can register namespace for a lg file. Also, they can register some templates as a expression functions, which can be applied easily in dialog. 
Here is an example to use these options,
> !# @registerNamespace = foo
> !# @registerAsFunctions = template1, template2

After registration,  in a dialog file, such usage `"property": "=foo.template1()"` is enabled, foo.template1() just behaves like a customized function.

The limitation is this kind of 'LG template function' can not read any scope from dialogStateManager, which is the same as prebuilt functions. 

For example, 
we have a memory in dialogStateManager:
`user.tasks: [ "car", "washing", "food", "laundry" ]`
if you to join the list of user.tasks, you have to define the LG file as 
```
> !# @registerNamespace = foo
> !# @registerAsFunctions = joinList

#joinList(tasks)
- ${join(tasks, " ,")} 

```

Then call the function as 
```
=foo.joinList(user.tasks)
```

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 1]
```diff
@@ -141,6 +141,8 @@ public class Templates : List<Template>
         /// </value>
         public EvaluationOptions LgOptions => new EvaluationOptions(Options);
 
+        public string NameSpace => ExtractNameSpace(Options);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:144]: >NameSpace [](start = 22, length = 9)

Namespace is one word, no capitol S

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 2]
```diff
@@ -326,6 +338,19 @@ public override bool Equals(object obj)
 
         public override int GetHashCode() => (Id, Content).GetHashCode();
 
+        public void InjectToExprssionFunction()
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:341]: >InjectToExprssionFunction [](start = 20, length = 25)

typo in expression name (missing E)
I would call this RegisterExpressionFunction(), that matches other functions we have called Register

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 3]
```diff
@@ -395,5 +420,66 @@ private void CheckErrors()
                 }
             }
         }
+
+        private string ExtractNameSpace(IList<string> optionStrList)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:424]: >optionStrList [](start = 54, length = 13)

how about "options"

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 4]
```diff
@@ -395,5 +420,66 @@ private void CheckErrors()
...
+        {
+            string result = null;
+            foreach (var optionStr in optionStrList)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:427]: >optionStr [](start = 25, length = 9)

option

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 5]
```diff
@@ -395,5 +420,66 @@ private void CheckErrors()
...
+                else
+                {
+                    throw new Exception("namespace is required!");
+                }
+            }
+
+            return result;
+        }
+
+        private IList<string> GetGlobalFunctionTable(IList<string> optionStrList)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:456]: options

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 6]
```diff
@@ -395,5 +420,66 @@ private void CheckErrors()
...
+        {
+            var result = new List<string>();
+            foreach (var optionStr in optionStrList)
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:459]: We abadoned hungarian notation about 20 years ago, no Str postfix

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿> !# @strict = false
+> !# @registerNamespace = foo
```
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg:2]: >registerNamespace [](start = 6, length = 17)

I would name this "Namespace"
```
!# Namespace = foo
```

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+> !# @registerAsFunctions = bar, cool
```
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg:3]: >registerAsFunctions [](start = 6, length = 19)

I would name this "Exports"or "Functions"
```
!# @Exports = bar, cool
!# @Functions = bar, cool
```

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 1 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: fixed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 2 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: fixed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 4 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: changed to option

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 5 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: fixed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 6 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: lesson learned, changed to option

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg (hunk 7 above)
Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg]: changed to Namespace

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg (hunk 8 above)
Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/Examples/InjectionTest/inject.lg]: Now is Exports

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 3 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: I changed it to options.

**Comment by xieofxie** [comment]
Will this apply for the dialogs under dialog having this lg or all dialogs? Thanks

**Comment by cosmicshuai** [comment]
@xieofxie, all dialogs

**Comment by xieofxie** [comment]
> @xieofxie, all dialogs

So we will load all lgs when bot is starting?

**Comment by cosmicshuai** [comment]
> > @xieofxie, all dialogs
> 
> So we will load all lgs when bot is starting?

Yes,  all LGs are loaded when bot starting. All exported templates can be used as a prebuilt functions under all dialogs.

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs [hunk 9]
```diff
@@ -1250,6 +1250,22 @@ public void TestCustomFunction2()
             Assert.AreEqual(6.0, evaled);
         }
 
+        [TestMethod]
+        public void TestInjectLG()
+        {
+            var templates = Templates.ParseFile(GetExampleFilePath("./InjectionTest/inject.lg"));
+            
+            var evaled = templates.EvaluateText("${foo.bar()}");
```
Comment (Danieladu) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs:1258]: maybe use Expression to test it is more suitable.
```
Expression.Parse("foo.bar()").TryEvaluate(xxx)
```

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs [hunk 10]
```diff
@@ -1250,6 +1250,22 @@ public void TestCustomFunction2()
...
+            
+            Assert.AreEqual("3", evaled.ToString());
+
+            evaled = templates.EvaluateText("${foo.cool(2)}");
+            Assert.AreEqual("3", evaled.ToString());
+
+            evaled = templates.EvaluateText("${common.looking()}");
+            Assert.AreEqual("John", evaled);
+        }
+
```
Comment (Danieladu) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs:1268]: Add more nested tests

Comment (Danieladu) [inline tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs:1268]: Add more Exception tests. 1.parameters is not match and so on

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs [hunk 11]
```diff
@@ -298,6 +292,26 @@ public object ConstructScope(string inputTemplateName, List<object> args)
             return new CustomizedMemory(memory.GlobalMemory, new SimpleObjectMemory(newScope));
         }
 
+        // Validator for template(...)
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs:295]: Please revert the code block if there is no change on it.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 12]
```diff
@@ -395,5 +435,66 @@ private void CheckErrors()
                 }
             }
         }
+
+        private string ExtractNameSpace(IList<string> options)
+        {
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:440]: We have tooo many operators on the split of "=" to get the key and value.
it is better to extract a new method `KeyValuePair<string, string> ExtractKeyValueOptions(string keyName)`

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 13]
```diff
@@ -151,7 +162,17 @@ public class Templates : List<Template>
         public static Templates ParseFile(
             string filePath,
             ImportResolverDelegate importResolver = null,
-            ExpressionParser expressionParser = null) => TemplatesParser.ParseFile(filePath, importResolver, expressionParser);
+            ExpressionParser expressionParser = null)
+        {
+            var curTemplate = TemplatesParser.ParseFile(filePath, importResolver, expressionParser);
+            curTemplate.InjectToExpressionFunction();
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:168]: curTemplates

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs [hunk 14]
```diff
@@ -151,7 +162,17 @@ public class Templates : List<Template>
...
+            foreach (var refer in curTemplate.References)
```
Comment (Danieladu) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:169]: Personally we should move it into the constructor. It is the behavior of templates initialization, but not only the behavior of ParseFile.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs (hunk 11 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs]: reverted

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs (hunk 10 above)
Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs:1283]: make sense, added exception tests

Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs:1283]: added

File: tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs (hunk 9 above)
Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.LanguageGeneration.Tests/TemplatesTest.cs]: That's the right way to test it. Thanks

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 12 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs:440]: refactor to a new method

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 13 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: removed

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs (hunk 14 above)
Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.LanguageGeneration/Templates.cs]: I tested the constructor. Found that, the injection should add to both constructor and parseFile (parseText also).

**Comment by boydc2014** [comment]
@cosmicshuai let's make sure our options are not case-sensitive before we take this in

**Pull request merged**